Support BlockMask pytree registration

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170088


Internal use case has mask_fn that gets defined inside another function. This trips our pytree spec extraction logic as dynamo fails to understand nestedFucntionVariable as python constant. This PR implements that. Also fixes a silly bug in export which should have caught this error earlier. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

tugsbayasgalan committed:
Support BlockMask pytree registration

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support BlockMask pytree registration"



There is a limitation that we can't reliably pytree mask_mod function. But i am not sure how we can get around it without resorting to hacks. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1579,6 +1579,9 @@ def __init__(
         return []
 
+    def as_python_constant(self):
+        return self.get_function()
```
What happens if user code causes this function to fail? Will we error out or graph break?

tugsbayasgalan committed:
Update on "Support BlockMask pytree registration"



Internal use case has mask_fn that gets defined inside another function. This trips our pytree spec extraction logic as dynamo fails to understand nestedFucntionVariable as python constant. This PR implements that. Also fixes a silly bug in export which should have caught this error earlier. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support BlockMask pytree registration"



Internal use case has mask_fn that gets defined inside another function. This trips our pytree spec extraction logic as dynamo fails to understand nestedFucntionVariable as python constant. This PR implements that. Also fixes a silly bug in export which should have caught this error earlier. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support BlockMask pytree registration"



Internal use case has mask_fn that gets defined inside another function. This trips our pytree spec extraction logic as dynamo fails to understand nestedFucntionVariable as python constant. This PR implements that. Also fixes a silly bug in export which should have caught this error earlier. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support BlockMask pytree registration"



Internal use case has mask_fn that gets defined inside another function. This trips our pytree spec extraction logic as dynamo fails to understand nestedFucntionVariable as python constant. This PR implements that. Also fixes a silly bug in export which should have caught this error earlier. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support BlockMask pytree registration"



Internal use case has mask_fn that gets defined inside another function. This trips our pytree spec extraction logic as dynamo fails to understand nestedFucntionVariable as python constant. This PR implements that. Also fixes a silly bug in export which should have caught this error earlier. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support BlockMask pytree registration"



Internal use case has mask_fn that gets defined inside another function. This trips our pytree spec extraction logic as dynamo fails to understand nestedFucntionVariable as python constant. This PR implements that. Also fixes a silly bug in export which should have caught this error earlier. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 reviewed (approved):
some comments on error message o/w lgtm

tugsbayasgalan committed:
Update on "Support BlockMask pytree registration"



Internal use case has mask_fn that gets defined inside another function. This trips our pytree spec extraction logic as dynamo fails to understand nestedFucntionVariable as python constant. This PR implements that. Also fixes a silly bug in export which should have caught this error earlier. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

drisspg commented on test/export/test_experimental.py:
```diff
@@ -161,6 +161,104 @@ def forward(self, p_linear_weight, p_linear_bias, c_lifted_tensor_0, x):
+        self.assertEqual(out_eager, out_compiled)
+        # they are not same function because dynamo reconstructs them
+        self.assertEqual(
+            mask_eager.mask_mod(B, H, Q_LEN, KV_LEN),
```
I feel like this isnt doing anything we should acutally run over teh full range and make sure all values are equal

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan commented:
@pytorchbot merge -f "Unrelated failures"
